webapp/store/selectors/nav.js

```diff
@@ -0,0 +1,189 @@
+import assert from 'assert';
+import { createSelector } from 'reselect';
+import { plugins } from '@bpanel/bpanel-utils';
+
+const { comparePlugins } = plugins;
+
+const getMetadataList = state => Object.values(state.pluginMetadata);
+const getSidebarItems = state => state.nav.sidebar;
+
+/* Sort plugin metadata (by order and name) and insert and
+ * sort children after parents
+ * @param {Array{}} pluginMeta - unsorted array of plugin Metadata
+ * @returns {Array{}} plugins - returns an array of sorted metadata objects
+ */
+export const sortPluginMetadata = (pluginMeta = []) => {
+  assert(Array.isArray(pluginMeta), 'Must pass array of metadata');
+  const subItems = new Map();
+
```

Comment:
Yeah, I'm not in love with the way a lot of this works, but couldn't think of how to improve it since it has this reliance on the state. I'm open to suggestions though. It shouldn't have any side effects however as it takes in something from the state and returns a new array.

Reply:
By side effects I mean inside of `filter` there is a reference to a global `Map` which is then iterated over with `forEach` and mutating an array. It just feels like there is a high surface area for bugs to appear